tinygrad/codegen/uopgraph.py

```diff
@@ -477,9 +481,29 @@ def delete_redundant_gates(root:UOp) -> Optional[UOp]:
   def find_gate(x:UOp) -> Optional[UOp]:
     if x.op is UOps.IF: return x
     return next((ret for s in x.src if (ret:=find_gate(s)) is not None), None)
-  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3]: return None
+  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3]:
+    if (sub_gate := find_gate(root.src[2])) is not None and sub_gate.src[0] is root.src[-1].src[0]:
+      return UOp(UOps.STORE, root.dtype, root.src[:3], root.arg)
+    return None
   return UOp(UOps.STORE, root.dtype, root.src[:3], root.arg)
 
+def merge_gates(sink:UOp) -> Optional[UOp]:
+  @functools.lru_cache(None)
+  def has_range(x:UOp) -> bool: return any(s2p.op is UOps.RANGE for s2p in x.sparents)
+  if not any(x.op is UOps.STORE and len(x.src) == 4 and x.src[-1].op is UOps.IF for x in sink.src): return None
```

Comment:
will STORE ever have len(x.src) == 4 and not be IF here? I think all gates should have been rewritten already, aren't they?

Reply:
Oh ya good catch, I guess I was just being extra cautious. Will update now

Reply:
"extra caution" is the wrong way to think about the uopgraph codebase.
The pattern matcher must be deterministic and have clear, localized behavior.